### ISSUE TITLE
pca_transform_result.transformedData seems to leak memory

### ISSUE THREAD

**Issue by oleksandr-pavlyk** [IntelPython/daal4py#92]
@fschlimb 

Consider the following script:

```python
import numpy as np
import daal4py
import os
import psutil

X = np.zeros((10**6, 100))
U = np.full((10, 100), 0.3, dtype='d')

n_iters = 200
proc = psutil.Process(os.getpid())
rss0 = proc.memory_info().rss

for _ in range(n_iters):
    res = daal4py.pca_transform(fptype='double').compute(
        X, U, dict()
    )
    Y = res.transformedData
    del res
    del Y
rss1 = proc.memory_info().rss

print( (rss1 - rss0) / X.nbytes )
```

Running it, prints out a number close to the value of `n_iter` for any of its value large enough, suggesting that memory is leaked within that loop.

Commenting out lines `Y = res.transformedData` and `del Y`, the value printed out becomes independent of the value of `n_iter`.

Having run Intel(R) Inspector `--collect mi1`, I see the following 

<details><summary><b>Header of output of `inspxe-cl --report problems`</b></summary>
<p>

```text
(vanilla) [09:51:22 ansatskl1004 pca_transform_memory_leak]$ inspxe-cl --report problems
P1: Error: Memory leak: New
 P1.147: Error: Memory leak: 32 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x32023d: Error X147: Allocation site: Function [Unknown]: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P2: Error: Memory leak: New
 P2.148: Error: Memory leak: 32 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3c15cb: Error X148: Allocation site: Function [Unknown]: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P3: Error: Memory leak: New
 P3.149: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3c1ded: Error X149: Allocation site: Function make_nt: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P4: Error: Memory leak: New
 P4.150: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3c9a1f: Error X150: Allocation site: Function make_dnt: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P5: Error: Memory leak: New
 P5.151: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3d7b26: Error X151: Allocation site: Function SharedPtr<unsigned char, daal::services::interface1::ServiceDeleter>: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P6: Error: Memory leak: New
 P6.152: Error: Memory leak: 48 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3dbda6: Error X152: Allocation site: Function NumericTable: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P7: Error: Memory leak: New
 P7.153: Error: Memory leak: 48 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3dbe00: Error X153: Allocation site: Function NumericTable: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P8: Error: Memory leak: New
 P8.154: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3dc0a4: Error X154: Allocation site: Function NumericTable: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P9: Error: Memory leak: New
 P9.155: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3dc0fc: Error X155: Allocation site: Function NumericTable: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P10: Error: Memory leak: New
 P10.156: Error: Memory leak: 24 Bytes: New
  _daal4py.cpython-36m-x86_64-linux-gnu.so!0x3f3dae: Error X156: Allocation site: Function create: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/_daal4py.cpython-36m-x86_64-linux-gnu.so

P11: Error: Memory leak: New
 P11.143: Error: Memory leak: 72 Bytes: New
  libdaal_core.so!0x1cd7c8b: Error X143: Allocation site: Function Argument: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/../../libdaal_core.so

P12: Error: Memory leak: New
 P12.144: Error: Memory leak: 56 Bytes: New
  libdaal_core.so!0x210356c: Error X144: Allocation site: Function Atomic: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/../../libdaal_core.so

P13: Error: Memory leak: New
 P13.160: Error: Memory leak: 80013344 Bytes: New
  libdaal_core.so!0x5acff49: Error X160: Allocation site: Function fpk_serv_malloc: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/../../libdaal_core.so

P14: Error: Memory leak: New
 P14.146: Error: Memory leak: 8 Bytes: New
  libdaal_core.so!0x5ad0122: Error X146: Allocation site: Function fpk_serv_malloc: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/lib/python3.6/site-packages/../../libdaal_core.so

P15: Error: Memory leak: New
 P15.157: Error: Memory leak: 185127 Bytes: New
  /tmp/build/80754af9/python_1546130271559/work/Objects/obmalloc.c(1446): Error X157: Allocation site: Function _PyObject_Alloc.isra.0: Module /localdisk/work/opavlyk/miniconda3_cb3/envs/vanilla/bin/python3.6
```
</p>
</details>

**Comment by fschlimb** [comment]
@Alexander-Makaryev

**Comment by oleksandr-pavlyk** [comment]
Here is my understanding of what is going on.

```
    @property
    def transformedData(self):
        '''
        transformedData
        :type: Numpy array
        '''
        if not is_valid_ptrptr(self.c_ptr):
            raise ValueError("Pointer to DAAL entity is NULL")
        res = get_pca_transform_result_transformedData(self.c_ptr)
        return <object>make_nda(res)
```

where `self.c_ptr` is declared as ``cdef pca_transform_ResultPtr * c_ptr``.

`get_pca_transform_result_transformedData` is 

```
extern "C" daal::data_management::NumericTablePtr * 
  get_pca_transform_result_transformedData(
    daal::algorithms::pca::transform::ResultPtr * obj_)
{
    return RAW< daal::data_management::NumericTablePtr >()(
          (*obj_)->get(daal::algorithms::pca::transform::transformedData)
    );
}
```

The `RAW` is a struct with `operator()` defined:

```
template< typename T >
struct RAW
{
    typedef T TYPE;
    const TYPE operator()(const T & o) {return o;}
};

template< typename T >
struct RAW< daal::services::SharedPtr< T > >
{
    typedef daal::services::SharedPtr< T > * TYPE;
    TYPE operator()(daal::services::SharedPtr< T > o) {
        return new daal::services::SharedPtr< T >(o);
    }
};
```

essentially, we increment the reference counter of the shared pointer `o` by calling the constructor, 
but never decrement it. 

The simple C++ reproducer is 

```c++
#include "daal.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;

template< typename T >
struct RAW
{
    typedef T TYPE;
    const TYPE operator()(const T & o) {return o;}
};

template< typename T >
struct RAW< daal::services::SharedPtr< T > >
{
    typedef daal::services::SharedPtr< T > * TYPE;
    TYPE operator()(const daal::services::SharedPtr< T > & o) { return new daal::services::SharedPtr< T >(o); }
};

int
main(void) {
    const size_t nSamples = 1000000;
    const size_t nFeatures = 100;
    const size_t nComponents = 10;

    double *Xmat = new double[nSamples * nFeatures];
    for(size_t i = 0; i <  nSamples * nFeatures; i++)
            Xmat[i]  = 0.0;

    double *Umat = new double[nComponents * nFeatures];
    for(size_t i = 0; i <  nFeatures * nComponents; i++)
            Umat[i]  = 1.0;


    services::SharedPtr<HomogenNumericTable<double> > X_hnt = HomogenNumericTable<double>::create(Xmat, nFeatures, nSamples);
    services::SharedPtr<HomogenNumericTable<double> > U_hnt = HomogenNumericTable<double>::create(Umat, nFeatures, nComponents);

    for (size_t it = 0; it < 2500; it++) {
        pca::transform::Batch<double> pcaTransform(nComponents);
        pcaTransform.input.set(pca::transform::data, X_hnt);
        pcaTransform.input.set(pca::transform::eigenvectors, U_hnt);

        pcaTransform.compute();

        pca::transform::ResultPtr pcaTransformResult = pcaTransform.getResult();
        // Use of RAW causes memory leak. 
        // Comment the next line out, and uncomment the one below it to fix the memory leak.
        data_management::NumericTablePtr * R_raw = RAW< data_management::NumericTablePtr >()(
            (pcaTransformResult)->get(pca::transform::transformedData));
        // NumericTablePtr R_hnt = pcaTransformResult->get(pca::transform::transformedData);                                                                                                             
    }

    delete[] Xmat;
    delete[] Umat;
}
```

**Comment by fschlimb** [comment]
@Alexander-Makaryev Sasha provided a partial fix on branch fix-mem-leak

**Comment by fschlimb** [comment]
I stepped through it in a debugger and I can't see any issue that could explain inspector's report. Deletion of shared-ptrs and DAAL objects (algo and result) happen as intended.

**Comment by fschlimb** [comment]
Fixed with #94 and #96
